### ISSUE TITLE
return of auto-increment expression mistranslated

### ISSUE THREAD

**Issue by jcburley** [elliotchance/c2go#787]
Whether one has this:

```
int
inc(void)
{
  static int counter;
  return ++counter;
}
```

Or this:

```
int
inc(void)
{
  static int counter;
  return counter++;
}
```

The resulting translation contains:

```
func inc() int32 {
	var counter int32
	return counter += 1
}
```

And this does not seem to be accepted by **go**:

```
$ go build return-postinc.go
# command-line-arguments
./return-postinc.go:14:17: syntax error: unexpected += at end of statement
$
```

**Comment by elliotchance** [comment]
Ah yes, this has popped up once before. It's because it has to be wrapped (a little ugly) in a closure, like:

```go
return func () int { counter += 1; return counter }()
```

To stop this from happening everywhere and making the code unreadable it tries to decide if the closure is needed, but the nesting of expressions can get quite complicated. In this case it has made the wrong decision.

Perhaps the easiest way to debug is to `git bisect` back until the code is working.

**Comment by jcburley** [comment]
FYI, I just now noticed that postincrement in a `while` seems to be handled properly:

```
  while (i--) {
```

This becomes:

```
	for func() int32 {
		defer func() {
			i -= 1
		}()
		return i
	}() != 0 {
```

Presumably the `return` transpilation could do something similar?

**Comment by elliotchance** [comment]
Yes, as I said the mechanism to determine if the expression can remain inline or must be translated is not working for that return expression.
